container/rules/docker_toolchains.bzl

```diff
@@ -269,6 +234,81 @@ def _toolchain_container_impl(ctx):
+
+    debian_output = _debian_pkg_tar_impl(ctx, download_pkgs_output_tar, download_pkgs_output_script)
+
+    return _language_tool_layer_impl(
+        ctx,
+        symlinks = debian_output.providers[0].symlinks,
+        env = debian_output.providers[0].env,
+        tars = debian_output.providers[0].tars,
+        files = debian_output.providers[0].files,
+        packages = debian_output.providers[0].packages,
+        additional_repos = debian_output.providers[0].additional_repos,
+        keys = debian_output.providers[0].keys,
+        installables_tars = debian_output.providers[0].installables_tars,
+        installation_cleanup_commands = debian_output.providers[0].installation_cleanup_commands,
+    )
+
+toolchain_container_ = rule(
+    attrs = language_tool_layer_attrs + {
+        "language_layers": attr.label_list(),
+    },
+    executable = True,
+    outputs = _container.image.outputs,
+    implementation = _toolchain_container_impl,
+)
+
+def toolchain_container(**kwargs):
+    """Composes multiple language_tool_layers into a single resulting image.
+
+    Args:
+      Same args as https://github.com/bazelbuild/rules_docker#container_image-1
+      minus:
+        debs: debian packages should be listed in 'packages', or be included in
+              'installables_tar' as .deb files.
+      plus:
+        language_layers: a list of language_tool_layer.
+        installables_tar: a tar of debian packages to install in the base image.
+        packages: list of packages to fetch and install in the base image.
+        additional_repos: list of additional debian package repos to use,
+          in sources.list format.
+        keys: list of labels of additional gpg keys to use while downloading
+          packages.
+        installation_cleanup_commands: cleanup commands to run after package
+          installation.
+
+    If 'installables_tar' is specified in the 'toolchain_container' rule, then
+    'packages' or 'installables_tar' specified in any of the 'language_layers'
+    passed to this 'toolchain_container' rule will be ignored.
+
+    Experimental rule.
+    """
+
+    _input_validation(kwargs)
+
+    toolchain_container_(**kwargs)
+
+def _debian_pkg_tar_impl(
+        ctx,
+        download_pkgs_output_tar = None,
+        download_pkgs_output_script = None):
+    """Implementation for the debian_pkg_tar rule.
+
+    debian_pkg_tar rule produces a tarball with all debian packages declared plus
+    those in the language_tool_layer(s) this rule depends on.
+
+    Args:
+      ctx: ctx as the same as for container_image + list of language_tool_layer(s)
+      https://github.com/bazelbuild/rules_docker#container_image
+      download_pkgs_output_tar: output tar file generated by download_pkgs rule to
+      override default output_tar name
+      download_pkgs_output_script: output script generated by download_pkgs rule to
+      override default output_script name
+    """
+
```

Comment:
_debian_pkg_tar_impl has no business doing anything related to env, symlinks, files, installation_cleanup_commands  or tars (or even the installables_tar attr). It should only deal with packages, additional_repos and keys. This rule should only receive base, and language_tool_layer(s), and that's about it.

Reply:
Will refactor the new rule in the next patch.